mitiq/shadows/shadows.py

```diff
@@ -0,0 +1,206 @@
+
+import cirq
+import numpy as np
+from numpy.typing import NDArray
+
+import mitiq
+from mitiq import MeasurementResult
+from mitiq.shadows.classical_postprocessing import (
+    expectation_estimation_shadow,
+    get_pauli_fidelities,
+    shadow_state_reconstruction,
+)
+from mitiq.shadows.quantum_processing import random_pauli_measurement
+
+
+def pauli_twirling_calibrate(
+    k_calibration: int = 1,
+    locality: Optional[int] = None,
+    zero_state_shadow_outcomes: Optional[Tuple[List[str], List[str]]] = None,
+    qubits: Optional[List[cirq.Qid]] = None,
+    executor: Optional[Callable[[cirq.Circuit], MeasurementResult]] = None,
+    num_total_measurements_calibration: Optional[int] = 20000,
+) -> Dict[str, complex]:
+    r"""
+    This function returns the dictionary of the median of means estimation
+    of Pauli fidelities: :math:`\{`"b": :math:`f_{b}\}_{b\in\{0,1\}^n}`.
+    The number of :math:`f_b` is :math:`2^n`, or :math:`\sum_{i=1}^d C_n^i` if
+    the locality :math:`d` is given.
+
+    In the notation of arXiv:2011.09636, this function estimates the
+    coefficient :math:`f_b`, which are expansion coefficients of the twirled
+    channel :math:`\mathcal{M}=\sum_b f_b\Pi_b`.
+
+    In practice, the output of this function can be used as calibration data
+    for performing the classical shadows protocol in a way which is more
+    robust to noise.
+
+    Args:
+        k_calibration: Number of groups of "median of means" used to solve for
+            Pauli fidelity.
+        locality: The locality of the operator, whose expectation value is
+            going to be estimated by the classical shadow. e.g. if operator is
+            Ising model Hamiltonian with nearest neighbour interaction, then
+            locality = 2.
+        zero_state_shadow_outcomes: The output of function
+            :func:`shadow_quantum_processing` of zero calibrate state.
+        qubits: The qubits to measure, needs to specify when the
+            `zero_state_shadow_outcomes` is None.
+        executor: The function to use to do quantum measurement, must be same
+            as executor in `shadow_quantum_processing`. Needs to specify when
+            the `zero_state_shadow_outcomes` is None.
+        num_total_measurements_calibration: Number of shots per group of
+            "median of means" used for calibration. Needs to specify when
+            the `zero_state_shadow_outcomes` is None.
+
+    Returns:
+        A dictionary containing the calibration outcomes.
+    """
+    if zero_state_shadow_outcomes is None:
+        if qubits is None:
+            raise ValueError("qubits should not be None.")
+        if executor is None:
+            raise ValueError("executor should not be None.")
+        if num_total_measurements_calibration is None:
+            raise ValueError(
+                "num_total_measurements_calibration should not be None."
+            )
```

Comment:
Wondering if these should be `TypeError`s?

Reply:
btw, I have removed 
```
# Check that the result is a dictionary
assert isinstance(result, dict)

```
as you suggested!

Reply:
BTW in https://github.com/unitaryfund/mitiq/pull/1921#pullrequestreview-1577943726 I did not recommend you remove the `assert` statement, only the comments that duplicate the code.